### PR TITLE
CIP-0101 | Integration of `keccak256` into Plutus

### PR DESCRIPTION
**Pull request by serejke**
This CIP proposes to add `keccak256` hash function to Plutus to unlock the potential of cross-chain integrations with Ethereum.

----------------

[Rendered proposal in branch](https://github.com/serejke/CIPs/blob/CIP-keccak256/CIP-%3F%3F%3F%3F/README.md)

### CODE REVIEW

**Comment by iquerejeta** [comment]
Thanks for opening this! Given that the main motivation of introducing ECDSA over SECP256k1 in plutus (#250) was to be compatible with Ethereum, I believe that including keccak256 as a built-in functions is a _must_. Otherwise, we would not achieve our initial goal of enabling cross-chain applications.  

As it is mentioned in the CIP, introducing such a built-in function should be pretty straightforward. The primitive is already part of cardano-base, and would only require one additional built-in function in Plutus. 

I fully support this going forward.

**Review by rphair — COMMENTED** [review]
@michaelpj for the Plutus implementors to accept this, wouldn't the CIP need to include the level of detail seen in CIP-0049? (maybe not necessary since "already available in cardano-base"?) https://github.com/cardano-foundation/CIPs/blob/master/CIP-0049/README.md

@serejke ... you might have used an earlier CIP as a model but we now have a more rigorous form for CIP documents.  Can you edit your header & headings so they're consistent with this specification? https://github.com/cardano-foundation/CIPs/blob/master/CIP-0001/README.md

e.g. there's no _Path to Proposed_ since your document will be submitted with `Proposed` status (and should say so in the header).

Hopefully we can address these formatting & content issues in the next 2 weeks, when we've planned for Plutus representatives to be present at our next CIP meeting... I've marked it for Review in the agenda of CIP meeting # 69 after we introduce it at today's meeting (45 minutes from now).

(p.s. I see some duplication with @Ryun1 review as I was writing this)

**Comment by rphair** [comment]
Thanks @iquerejeta for coming to discuss this at the CIP editors' meeting today.  Regarding the lack of technical detail in the document compared to CIP-0049 we noted that this is just adding a hashing function rather than a full signature scheme.  Even so, we agreed that something should be added to the CIP text to explain this.

Also thanks @iquerejeta for agreeing to coordinate with @serejke about updating the CIP document format according to the new specification (CIP-0001) introduced in Q4 2022.  Hopefully these changes can be made before a more rigorous review in the CIP meeting planned 2 weeks from now.

**Comment by iquerejeta** [comment]
Yes! These changes will be addressed before the review meeting planned 2 weeks from now 👍 Thanks for addressing this CIP in today's session.

**Comment by hai-kreate** [comment]
This addition would 100% push interoperability forward. 
We will utilize it the moment it is available :).

**Comment by serejke** [comment]
Thanks @rphair for the valuable comments on the content/formatting of this CIP. I've tried to address all the issues: fixed the preamble headers, elaborated on the Specification. Look forward to the CIP editors' opinion.

File: CIP-????/README.md
Comment (rphair) [inline CIP-????/README.md:20]:
```suggestion
## Motivation: why is this CIP necessary?
```
This is a standard title which helps readers & implementors also understand what's being answered here (even if authors understand it) 😎

File: CIP-????/README.md
Comment (rphair) [inline CIP-????/README.md:47]:
```suggestion
## Rationale: how does this CIP achieve its goals?
```
(likewise with Motivation)

File: CIP-????/README.md
Comment (rphair) [inline CIP-????/README.md:53]:
```suggestion
### Backward Compatibility
```
There is no top-level section by this name (anymore) so demoting the heading will absorb this into `Rationale` which generally makes sense.  It might also make sense to drop this section there there are no backward compatibility issues here.

File: CIP-????/README.md
Comment (rphair) [inline CIP-????/README.md:57]: We are looking for a section here which discriminates between Acceptance and Implementation: https://github.com/cardano-foundation/CIPs/blob/master/.github/CIP-TEMPLATE.md#path-to-active

Sometimes these are trivial, e.g. "Acceptance is determined when the Plutus team(s) agree to do it" but please add whatever you can now about what actually needs to happen and who will be responsible for it.

When this comes up at the CIP editors' meeting (please attend if you can) I'll plan to mention this question along with the question of whether the overall level of detail is sufficient.

File: CIP-????/README.md
Comment (rphair) [inline CIP-????/README.md:59]: There is also no `References` section in our standard template: https://github.com/cardano-foundation/CIPs/blob/master/.github/CIP-TEMPLATE.md

Depending on the Markdown format you use, these might remain at the end of your document as footnotes.  Otherwise the reference links should be included inline within the document where they are referenced.

File: CIP-????/README.md
Comment (rphair) [inline CIP-????/README.md:63]:
```suggestion
- Ethereum standard on signing data: https://eips.ethereum.org/EIPS/eip-712

## Copyright
This CIP is licensed under [Apache-2.0][https://www.apache.org/licenses/LICENSE-2.0]. 
```
Finally our template includes a Copyright section at the end matches the one named in the document header: https://github.com/cardano-foundation/CIPs/blob/master/CIP-0001/README.md#licensing

**Review by rphair — COMMENTED** [review]
Thanks for the update @serejke ... the format is almost ready for review with just a few nitpicks & more specific recommendations about making it consistent with the CIP-0001 requirements.

I believe there is popular demand for the next CIP editors' meeting being Plutus themed, after more people get back from June holidays.  My own first criteria for review would be first to ensure that the Plutus implementors have enough detail in this CIP for it to meet their own requirements & I would endorse it if so.

File: CIP-????/README.md
Comment (rphair) [reply CIP-????/README.md:54]: p.s. @serejke I can see that an expansion of this section was underway in https://github.com/cardano-foundation/CIPs/pull/528 so we'll look forward to reviewing those additions here in this PR branch. 🤓

**Comment by rphair** [comment]
FYI @serejke @iquerejeta CIP editors are all in agreement about assigning a number to this proposal as an official candidate... though it's still awaiting review by us (and Plutus reps) & some editing by you 🤓

File: CIP-????/README.md
Comment (serejke) [reply CIP-????/README.md:54]: @rphair I have a multi-stage approval: first, in my fork by the more experienced colleague @iquerejeta :-), then pushing into this branch. We're going to update the PR very soon. Thanks for keeping an eye on it 🙏

Comment (serejke) [reply CIP-????/README.md]: Done! Look forward to your opinion

File: CIP-????/README.md
Comment (rphair) [inline CIP-????/README.md:52]:
```suggestion
in a node release and becomes available on the Mainnet.
```
also: the word "the" is optional here (not needed and usually not used)

**Review by rphair — COMMENTED** [review]
new Path to Active section looks good... just one spelling fix please 🤓

**Comment by rphair** [comment]
@serejke @iquerejeta this, with other Plutus items, is on the agenda for review early in the CIP editors' meeting Tuesday 04 July 16:00 UTC: https://hackmd.io/@cip-editors/69

**Comment by iquerejeta** [comment]
Great! Will be there 👍

**Review by KtorZ — COMMENTED** [review]
Following the process outlined by the Plutus team as part of [CIP-0035](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0035), I believe that this CIP still misses a few things:

The Specification MUST include:
- [x] ~~Names and types/kinds for the new functions or types.~~
- [ ] A source for the implementation (e.g. a library which can be linked against); or a generic description of the functionality which is implementable in any programming language.
- [ ] For new builtin functions: a costing function for the builtin function.

The Rationale MUST include:
- [ ] If an external implementation is provided: an argument that it satisfies the following non-exhaustive list of criteria:
  - It is trustworthy
  - It always terminates
  - It (or its Haskell bindings) never throw any exceptions
  - Its behaviour is predictable (e.g. does not have worst-case behaviour with much worse performance)
- [ ]  Discussion of how any measures and costing functions were determined.

The Acceptance Criteria MUST include:
- [ ] The ledger is updated to include new protocol parameters to control costing of the new builtins.

**edit: updated checklist** in https://github.com/cardano-foundation/CIPs/pull/524#pullrequestreview-1549528448
--- 

Regarding the implementation, I believe that anything which piggybacks on `libsodium` might be okay given that it's already used as a dependency for a couple of other things. Otherwise, [`cryptonite`](https://hackage.haskell.org/package/cryptonite) seems also like a good candidate to me as it supports Keccak as a hashing algorithm.

**Comment by iquerejeta** [comment]
The specification already mentions that we'd use cryptonite, and that the bindings are available in cardano-base. See [here](https://github.com/cardano-foundation/CIPs/pull/524/files#diff-fb18fdd6d900ee9d47482161e923dac78533555da2f57a8f7c3c06c0c8bd0611R35-R38). I guess this is not considered an external implementation, right?

We still have not costed the function. This is marked as acceptance criteria for being active.

**Comment by KtorZ** [comment]
> I guess this is not considered an external implementation, right?

Well. Tricky given Vincent is the main author indeed. Though the library being an external dependency, I think it still counts as an external implementation (as in, not something implemented by and under control of the Plutus team). 

> We still have not costed the function. This is marked as acceptance criteria for being active.

The Plutus team expects this as part of the specification in order to even be considered as `proposed` as far as I know.

**Review by michaelpj — COMMENTED** [review]
I think this is a reasonable thing to have. We mostly haven't done it because we lack _clear use cases_. This CIP also doesn't provide any! So at the moment it's not yet clear to me that this _is_ actually useful.

File: CIP-????/README.md
Comment (iquerejeta) [reply CIP-????/README.md:36]: We do have that as a low priority task. The reason why we want to move away from it is because we have cryptonite as a dependency for a single primitive (SHA3), not due to any security concerns. However, we are not planning on moving away from cryptonite anytime soon, so using it's keccak implementation is the easiest. 

If we ever make this change in the near future (moving away from cryptonite) there will be no effects on the interface or behaviour of the primitives, so there would not need to be a hard-fork.

File: CIP-????/README.md
Comment (iquerejeta) [reply CIP-????/README.md:20]: Sounds good. @hai-kreate, @serejke , could you add a brief description of the use cases you had in mind? A brief description of the bridge or whatever use case that requires compatibility with Ethereum.

Comment (serejke) [reply CIP-????/README.md:21]: I have added the use case we're actively work on ([Hyperlane](https://hyperlane.xyz/)). Hyperlane is a growing interchain communication protocol and we've figured out a way to _partially_ support Cardano by using its native crypto hash function (`blake2b`). Having the `keccak256` would provide additional guarantees (censorship resistance) and allow for cross-chain slashing (work in progress).

Comment (rphair) [reply CIP-????/README.md:21]: @michaelpj looks like fixed by https://github.com/cardano-foundation/CIPs/pull/524/commits/86ac8d20b7b72da354c022d726e33476eab758f2; please unresolve if not?

File: CIP-????/README.md
Comment (rphair) [reply CIP-????/README.md]: @michaelpj looks like fixed by https://github.com/cardano-foundation/CIPs/pull/524/commits/86ac8d20b7b72da354c022d726e33476eab758f2; please unresolve if not?

**Comment by rphair** [comment]
my recollection from the CIP meeting in the last hour (@iquerejeta @effectfully speaking, @michaelpj available on chat):

We noted that there are some particulars missing from the last 2 reviews above, but agreed to assign a number to this as a candidate & with the intention of merging as `Proposed` ASAP. (cc @KtorZ). The review included some discussion of the cost model called for in the reviews above.

File: CIP-????/README.md
Comment (iquerejeta) [inline CIP-????/README.md:65]:
```suggestion
* The function is benchmarked to assess its cost. As for other hash functions available in Plutus (blake2b and sha256), we expect the cost of keccak to be linear with respect to the size of the message.
```

Comment (iquerejeta) [reply CIP-????/README.md:65]: @rphair . I've suggested the comment that we expect the cost of the function to be linear wrt the message size as part of the AC. Do you think it should go somewhere else?

Comment (rphair) [reply CIP-????/README.md:65]: @iquerejeta it looks sufficient to me there.  @michaelpj I don't know if a more detailed cost analysis is required in the CIP itself as per cip35... I was struggling to hold onto any concrete requirements from the meeting so maybe @KtorZ @SebastienGllmt could better answer this.

File: CIP-????/README.md
Comment (hai-kreate) [reply CIP-????/README.md:21]: If it helps, I can propose a well-scoped use case that can be a baseline test. The TLDR is to post verifiable messages signed from the EVM world to Cardano. The go-to signature scheme over there is `ecdsa-secp256k1` which signs on fixed-size messages. It's then natural to sign on message hash to support arbitrary message size, and the go-to hash function over there is `keccak256`. 
1. With `keccak256` in Plutus, a minting policy issuing authentic tokens can hash the posted message (to an output datum) with `keccak256` then verify it against the signature posted in redeemer.
2. Without it, the minting policy would have to trust the messenger, or force EVM agents to sign on `blake2b` hashes or sign with `ed25519`, etc.

We're asking signers to sign on both `keccak256` and `blake2b` hashes at the moment for our project, but that wouldn't scale to a larger EVM ecosystem. Adding `keccak256` to Plutus, on the other hand, would open many doors for future integrations, like as @serejke outlined with Hyperlane.

I'm more than happy to work on this test case if accepted. We already have it written in Helios.

File: CIP-????/README.md
Comment (hai-kreate) [reply CIP-????/README.md]: Here's another: https://github.com/cardano-foundation/CIPs/pull/524#discussion_r1255491481 🙏.

**Comment by michaelpj** [comment]
The implementation is in progress [here](https://github.com/input-output-hk/plutus/pull/5431), so if we can get this CIP tidied up then I think we can safely merge it as Proposed.

File: CIP-????/README.md
Comment (rphair) [reply CIP-????/README.md]: Resolved in https://github.com/cardano-foundation/CIPs/pull/524/commits/b0dc15e3cc43c6486c3656ac621991aa9ea87d21.

File: CIP-????/README.md
Comment (rphair) [reply CIP-????/README.md:44]: noted... resolving these 2 `cryptonite` conversations to highlight practical blocks to merging the CIP.

File: CIP-????/README.md
Comment (rphair) [reply CIP-????/README.md:45]: noted... resolving these 2 `cryptonite` conversations (answered in https://github.com/cardano-foundation/CIPs/pull/524#discussion_r1252131805) to highlight practical blocks to merging the CIP.

**Review by rphair — COMMENTED** [review]
@serejke my reading of @michaelpj's latest https://github.com/cardano-foundation/CIPs/pull/524#issuecomment-1645368408 is that you need to confirm you've satisfied all the requirements in @KtorZ's https://github.com/cardano-foundation/CIPs/pull/524#pullrequestreview-1512910358.  

@iquerejeta if you can't modify this branch and can make these changes, just post them as a review and we can merge them here after confirming they are suitable.

I think the "costing function" requirement is satisified by confirming that it's linear in one of the more recent commits (?), so that box could be ticked right away (if you confirm these items, any editor can tick the boxes for you 🤓).

Once all that's done, I think we can merge this.

File: CIP-????/README.md
Comment (iquerejeta) [inline CIP-????/README.md:65]:
```suggestion
* The function is benchmarked to assess its cost. As for other hash functions available in Plutus (blake2b and sha256), we expect the cost of keccak to be linear with respect to the size of the message.
* The ledger is updated to include new protocol parameters to control costing of the new builtins.
```

File: CIP-????/README.md
Comment (iquerejeta) [inline CIP-????/README.md:57]:
```suggestion

The library, cryptonite, is not implemented by and under control of the Plutus team. However, 
* It is a library already used in the Cardano stack to expose SHA3, and can be considered as a trustworthy implementation. 
* The function does not throw any exceptions as hash functions are defined to work with any ByteString input. It does not expect a particular particular structure. 
* It's behaviour is predictable. As mentioned above, the cost of the function is linear with respect to the size of the message provided as input. This is the same behaviour that other hash functions exposed in plutus (blake, sha3) have.
```

**Review by iquerejeta — COMMENTED** [review]
I've added two comments to check a few boxes requested by @KtorZ . Costing will be done by the Plutus team, so we can update this CIP once we have costing done for keccak. I believe that other checkboxes can be marked. I didn't know how to argue that the function always terminates. Matthias, let me know if the rationale is complete in your opinion, or you think we should add something else.

**Review by rphair — COMMENTED** [review]
thanks @iquerejeta  I will be the bean-counter for this and keep track of remaining issues so we can merge it ASAP.  I've committed your suggestions in https://github.com/cardano-foundation/CIPs/pull/524#pullrequestreview-1542912382 which leaves these two items from the checklist in https://github.com/cardano-foundation/CIPs/pull/524#pullrequestreview-1512910358:

[x] an argument that ... It always terminates (https://github.com/cardano-foundation/CIPs/pull/524#issuecomment-1657220510)
[ ] Discussion of how any measures and costing functions were determined.

Since @KtorZ or another reviewer might help with the 1st item, it doesn't make sense to put this as `Waiting for Author` yet.  I'll keep watching this & if ready we might present it again @ the 01 August CIP meeting and merge it then. cc @serejke

**Comment by iquerejeta** [comment]
Ok. Regarding the argument that it always terminates, is it about the functionality (i.e. a hash function cannot fail as long as it receives a bytestring as input), or about the implementation itself?

**Comment by rphair** [comment]
@michaelpj wrote those requirements (https://github.com/cardano-foundation/CIPs/tree/master/CIP-0035#additions-to-the-plutus-core-builtins) so I think he's the best one to answer that 🤓

**Comment by michaelpj** [comment]
I think we're okay with the termination here. It would be very weird if a hash function didn't terminate. We do need to handle errors appropriately, but that's an implementation matter.

**Comment by rphair** [comment]
@serejke @iquerejeta (cc @michaelpj @kwxm) marking `Waiting for Author` due to this remaining unresolved item from https://github.com/cardano-foundation/CIPs/pull/524#pullrequestreview-1549528448:
- [ ] Discussion of how any measures and costing functions were determined.

**Comment by iquerejeta** [comment]
@michaelpj @kwxm , costing will be done (has been done?) by the Plutus team. What should we put in this section? Could you maybe point to some other CIP were we can copy the discussion on how costing will be determined?

**Comment by michaelpj** [comment]
Well, in many cases I would expect to have some determination like "these functions are constant-time because we know crypto operations are constant time", but in this case I think it's fine to say "costing functions were determined empirically by the plutus team", since that's what actually happened.

**Review by rphair — APPROVED** [review]
According to my understanding of https://github.com/cardano-foundation/CIPs/pull/524#issuecomment-1681916916, https://github.com/cardano-foundation/CIPs/pull/524/commits/644a12f45b200a348c46d14c0116fc20c8945bb6 fixes the last of the requirements so I believe this is good to go.

**Comment by kwxm** [comment]
> @michaelpj @kwxm , costing will be done (has been done?) by the Plutus team. What should we put in this section? Could you maybe point to some other CIP were we can copy the discussion on how costing will be determined?

We have costing benchmarks for the two new functions [here](https://github.com/input-output-hk/plutus/blob/250ba063b7cb8ecf9213d198a11d63c33ae39b17/plutus-core/cost-model/budgeting-bench/Benchmarks/Crypto.hs#L320).  Running these shows that the execution time is linear in the size of the input, as expected.  However, to get sensible coefficients for the CPU cost which are consistent with the costs for the other builtins we have to run all of the benchmarks for all of the builtins on a reference machine, and our previous machine disappeared without warning when Cicero was introduced.  We've been using an AWS instance for the benchmarking, but that's not very consistent.  We're currently awaiting a new physical machine and when that's available we'll recalibrate the entire Plutus Core cost model, including the new hash functions.  In the meantime I've reused the costing function for `sha3_256` for `keccak_256` since they're nearly the same algorithm, and I've re-used the costing function for `blake2b_256` for `blake2b_224` since the two functions are very similar, and the cost of the former should certainly be an upper bound for the cost of the latter.  

We don't anticipate any difficulty in getting the final costing figures once suitable hardware is available.

**Comment by kwxm** [comment]
> @michaelpj @kwxm , costing will be done (has been done?) by the Plutus team. What should we put in this section? Could you maybe point to some other CIP were we can copy the discussion on how costing will be determined?

We have costing benchmarks for the two new functions [here](https://github.com/input-output-hk/plutus/blob/250ba063b7cb8ecf9213d198a11d63c33ae39b17/plutus-core/cost-model/budgeting-bench/Benchmarks/Crypto.hs#L320).  Running these shows that the execution time is linear in the size of the input, as expected.  However, to get sensible coefficients for the CPU cost which are consistent with the costs for the other builtins we have to run all of the benchmarks for all of the builtins on a reference machine, and our previous machine disappeared without warning when Cicero was introduced.  We've been using an AWS instance for the benchmarking, but that's not very consistent.  We're currently awaiting a new physical machine and when that's available we'll recalibrate the entire Plutus Core cost model, including the new hash functions.  In the meantime I've reused the costing function for `sha3_256` for `keccak_256` since they're nearly the same algorithm, and I've re-used the costing function for `blake2b_256` for `blake2b_224` since the two functions are very similar, and the cost of the former should certainly be an upper bound for the cost of the latter.  

We don't anticipate any difficulty in getting the final costing figures once suitable hardware is available.  I think "The Plutus team determines the exact costing functions empirically" is a reasonable summary of all this.

**Comment by rphair** [comment]
@serejke this has passed our Last Check review at the CIP Editors' Meeting today, so please rename your CIP folder to `CIP-0101` and then we can provide the last required endorsing review and merge this.

**Review by Ryun1 — APPROVED** [review]

**Pull request merged**
